libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/SFHybridViewController.m

```diff
@@ -533,18 +581,124 @@ - (void) webView:(WKWebView *) webView decidePolicyForNavigationAction:(WKNaviga
+
+    // Local error page load.
+    if ([webView isEqual:self.errorPageUIWebView]) {
+        [self log:SFLogLevelDebug format:@"Local error page ('%@') is loading.", webView.request.URL.absoluteString];
+        return YES;
+    }
+
+    // Cordova web view load.
+    if ([webView isEqual:self.webView]) {
+
+        /*
+         * If the request is attempting to refresh an invalid session, take over
+         * the refresh process via the OAuth refresh flow in the container.
+         */
+        NSString *refreshUrl = [self isLoginRedirectUrl:webView.request.URL];
+        if (refreshUrl != nil) {
+            [self log:SFLogLevelWarning msg:@"Caught login redirect from session timeout. Reauthenticating."];
+
+            /*
+             * Reconfigure user agent. Basically this ensures that Cordova whitelisting won't apply to the
+             * WKWebView that hosts the login screen (important for SSO outside of Salesforce domains).
+             */
+            [SFSDKWebUtils configureUserAgent:[self sfHybridViewUserAgentString]];
+            [[SFAuthenticationManager sharedManager]
+             loginWithCompletion:^(SFOAuthInfo *authInfo) {
+
+                 // Reset the user agent back to Cordova.
+                 [self authenticationCompletion:refreshUrl authInfo:authInfo];
+             } failure:^(SFOAuthInfo *authInfo, NSError *error) {
+                 if ([self logoutOnInvalidCredentials:error]) {
+                     [self log:SFLogLevelError msg:@"Could not refresh expired session. Logging out."];
+                     [[SFAuthenticationManager sharedManager] logout];
+                 } else {
+
+                     // Error is not invalid credentials, or developer otherwise wants to handle it.
+                     [self loadErrorPageWithCode:error.code description:error.localizedDescription context:kErrorContextAuthExpiredSessionRefresh];
+                 }
+             }];
+            return NO;
+        }
+        NSURL* url = [request URL];
+        
+        /*
+         * Execute any commands queued with cordova.exec() on the JS side.
+         * The part of the URL after gap:// is irrelevant.
+         */
+        if ([[url scheme] isEqualToString:@"gap"]) {
+            [self.commandQueue fetchCommandsFromJs];
+            [self.commandQueue executePending];
+            return NO;
+        }
+
+        /*
+         * Give plugins the chance to handle the URL.
+         */
+        BOOL anyPluginsResponded = NO;
+        BOOL shouldAllowRequest = NO;
+        for (NSString* pluginName in self.pluginObjects) {
+            CDVPlugin* plugin = [self.pluginObjects objectForKey:pluginName];
+            SEL selector = NSSelectorFromString(@"shouldOverrideLoadWithRequest:navigationType:");
+            if ([plugin respondsToSelector:selector]) {
+                anyPluginsResponded = YES;
+                shouldAllowRequest = (((BOOL (*)(id, SEL, id, int)) objc_msgSend)(plugin, selector, request, navigationType));
```

Comment:
SalesforceHybridSDKTestApp also needs some changes/revert to get the unit tests working